// @umijs/bundler-vite/dist/dev.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dev = void 0;
const config_1 = require("./config/config");
const server_1 = require("./server/server");
const types_1 = require("./types");
async function dev(opts) {
    const viteConfig = await (0, config_1.getConfig)({
        cwd: opts.cwd,
        env: types_1.Env.development,
        entry: opts.entry,
        userConfig: opts.config,
        extraBabelPlugins: [
            ...(opts.beforeBabelPlugins || []),
            ...(opts.extraBabelPlugins || []),
        ],
        extraBabelPresets: [
            ...(opts.beforeBabelPresets || []),
            ...(opts.extraBabelPresets || []),
        ],
        modifyViteConfig: opts.modifyViteConfig,
    });
    await (0, server_1.createServer)({
        viteConfig,
        userConfig: opts.config,
        cwd: opts.cwd,
        port: opts.port,
        beforeMiddlewares: opts.beforeMiddlewares,
        afterMiddlewares: opts.afterMiddlewares,
        onDevCompileDone: opts.onDevCompileDone,
    });
}
exports.dev = dev;
