// @umijs/mfsu/dist/loader/esbuild.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.esbuildLoader = void 0;
const es_module_lexer_1 = require("@umijs/bundler-utils/compiled/es-module-lexer");
const esbuild_1 = require("@umijs/bundler-utils/compiled/esbuild");
const path_1 = require("path");
async function esbuildTranspiler(source) {
    var _a;
    const done = this.async();
    const options = this.getOptions();
    const { handler = [], implementation, ...otherOptions } = options;
    const transform = (implementation === null || implementation === void 0 ? void 0 : implementation.transform) || esbuild_1.transform;
    const filePath = this.resourcePath;
    const ext = (0, path_1.extname)(filePath).slice(1);
    const transformOptions = {
        ...otherOptions,
        target: (_a = options.target) !== null && _a !== void 0 ? _a : 'es2015',
        loader: ext !== null && ext !== void 0 ? ext : 'js',
        sourcemap: this.sourceMap,
        sourcefile: filePath,
    };
    try {
        let { code, map } = await transform(source, transformOptions);
        if (handler.length) {
            await es_module_lexer_1.init;
            handler.forEach((handle) => {
                const [imports, exports] = (0, es_module_lexer_1.parse)(code);
                code = handle({ code, imports, exports, filePath });
            });
        }
        done(null, code, map && JSON.parse(map));
    }
    catch (error) {
        done(error);
    }
}
exports.default = esbuildTranspiler;
exports.esbuildLoader = __filename;
