// @umijs/utils/dist/Generator/Generator.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = require("fs");
const path_1 = require("path");
const chalk_1 = __importDefault(require("../../compiled/chalk"));
const fs_extra_1 = __importDefault(require("../../compiled/fs-extra"));
const glob_1 = __importDefault(require("../../compiled/glob"));
const mustache_1 = __importDefault(require("../../compiled/mustache"));
const prompts_1 = __importDefault(require("../../compiled/prompts"));
class Generator {
    constructor({ baseDir, args }) {
        this.baseDir = baseDir;
        this.args = args;
        this.prompts = {};
    }
    async run() {
        const questions = this.prompting();
        this.prompts = await (0, prompts_1.default)(questions);
        await this.writing();
    }
    prompting() {
        return [];
    }
    async writing() { }
    copyTpl(opts) {
        const tpl = (0, fs_1.readFileSync)(opts.templatePath, 'utf-8');
        const content = mustache_1.default.render(tpl, opts.context);
        fs_extra_1.default.mkdirpSync((0, path_1.dirname)(opts.target));
        console.log(`${chalk_1.default.green('Write:')} ${(0, path_1.relative)(this.baseDir, opts.target)}`);
        (0, fs_1.writeFileSync)(opts.target, content, 'utf-8');
    }
    copyDirectory(opts) {
        const files = glob_1.default.sync('**/*', {
            cwd: opts.path,
            dot: true,
            ignore: ['**/node_modules/**'],
        });
        files.forEach((file) => {
            const absFile = (0, path_1.join)(opts.path, file);
            if ((0, fs_1.statSync)(absFile).isDirectory())
                return;
            if (file.endsWith('.tpl')) {
                this.copyTpl({
                    templatePath: absFile,
                    target: (0, path_1.join)(opts.target, file.replace(/\.tpl$/, '')),
                    context: opts.context,
                });
            }
            else {
                console.log(`${chalk_1.default.green('Copy: ')} ${file}`);
                const absTarget = (0, path_1.join)(opts.target, file);
                fs_extra_1.default.mkdirpSync((0, path_1.dirname)(absTarget));
                (0, fs_1.copyFileSync)(absFile, absTarget);
            }
        });
    }
}
exports.default = Generator;
