// @umijs/mfsu/dist/dep/getModuleExports.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getModuleExports = void 0;
const es_module_lexer_1 = require("@umijs/bundler-utils/compiled/es-module-lexer");
const esbuild_1 = require("@umijs/bundler-utils/compiled/esbuild");
const path_1 = require("path");
const getCJSExports_1 = require("./getCJSExports");
async function getModuleExports({ content, filePath, }) {
    // Support tsx and jsx
    if (filePath && /\.(tsx|jsx)$/.test(filePath)) {
        content = (await (0, esbuild_1.transform)(content, {
            sourcemap: false,
            sourcefile: filePath,
            format: 'esm',
            target: 'es6',
            loader: (0, path_1.extname)(filePath).slice(1),
        })).code;
    }
    await es_module_lexer_1.init;
    const [imports, exports] = (0, es_module_lexer_1.parse)(content);
    let isCJS = !imports.length && !exports.length;
    let cjsEsmExports = null;
    if (isCJS) {
        cjsEsmExports = (0, getCJSExports_1.getCJSExports)({ content });
        if (cjsEsmExports.includes('__esModule')) {
            isCJS = false;
        }
    }
    return {
        exports: cjsEsmExports || exports,
        isCJS,
    };
}
exports.getModuleExports = getModuleExports;
