// @umijs/lint/dist/config/eslint/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const recommended_1 = __importStar(require("./rules/recommended"));
require("./setup");
module.exports = {
    parser: '@babel/eslint-parser',
    plugins: ['react', 'react-hooks'],
    settings: {
        react: {
            version: 'detect',
        },
    },
    env: {
        browser: true,
        node: true,
        es2022: true,
        jest: true,
    },
    rules: recommended_1.default,
    overrides: [
        {
            parser: '@typescript-eslint/parser',
            plugins: ['@typescript-eslint/eslint-plugin'],
            files: ['**/*.{ts,tsx}'],
            rules: recommended_1.typescript,
        },
        {
            settings: {
                jest: {
                    version: 26,
                },
            },
            files: ['**/*.{test,spec,unit,e2e}.{ts,tsx,js,jsx}'],
            plugins: ['jest'],
            rules: recommended_1.jest,
        },
    ],
    parserOptions: {
        ecmaFeatures: {
            jsx: true,
        },
        babelOptions: {
            presets: [require.resolve('@umijs/babel-preset-umi')],
        },
        requireConfigFile: false,
    },
};
