// @umijs/preset-umi/dist/features/apiRoute/vercel/esbuild.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const esbuild_1 = __importDefault(require("@umijs/bundler-utils/compiled/esbuild"));
const path_1 = require("path");
const constants_1 = require("../constants");
const utils_1 = require("../utils");
// 将 API 路由的临时文件打包为 Vercel 的 Serverless Function 可以使用的格式
async function default_1(api, apiRoutes) {
    const apiRoutePaths = apiRoutes.map((r) => (0, path_1.join)(api.paths.absTmpPath, 'api', r.file));
    const pkg = require((0, path_1.join)(api.cwd, './package.json'));
    await esbuild_1.default.build({
        format: 'cjs',
        platform: 'node',
        bundle: true,
        entryPoints: [
            ...apiRoutePaths,
            (0, path_1.resolve)(api.paths.absTmpPath, 'api/_middlewares.ts'),
        ],
        outdir: (0, path_1.resolve)(api.paths.cwd, constants_1.OUTPUT_PATH),
        plugins: [(0, utils_1.esbuildIgnorePathPrefixPlugin)()],
        external: [...Object.keys(pkg.dependencies || {})],
    });
}
exports.default = default_1;
