// @umijs/preset-umi/dist/commands/dev/watch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.expandCSSPaths = exports.expandJSPaths = exports.unwatch = exports.createDebouncedHandler = exports.watch = exports.addUnWatch = void 0;
const utils_1 = require("@umijs/utils");
const unWatches = [];
function addUnWatch(unWatcher) {
    unWatches.push(unWatcher);
}
exports.addUnWatch = addUnWatch;
function watch(opts) {
    const watcher = utils_1.chokidar.watch(opts.path, {
        ...opts.watchOpts,
        ignoreInitial: true,
    });
    watcher.on('all', opts.onChange);
    if (opts.addToUnWatches) {
        addUnWatch(() => {
            watcher.close();
        });
    }
    return watcher;
}
exports.watch = watch;
function createDebouncedHandler(opts) {
    let timer = null;
    let files = [];
    return (event, path) => {
        if (timer) {
            clearTimeout(timer);
        }
        files.push({ event, path: (0, utils_1.winPath)(path) });
        timer = setTimeout(async () => {
            timer = null;
            await opts.onChange({ files });
            files = [];
        }, opts.timeout || 2000);
    };
}
exports.createDebouncedHandler = createDebouncedHandler;
function unwatch() {
    unWatches.forEach((unWatch) => unWatch());
}
exports.unwatch = unwatch;
function expandJSPaths(path) {
    return ['.js', '.jsx', '.ts', '.tsx'].map((ext) => {
        return `${path}${ext}`;
    });
}
exports.expandJSPaths = expandJSPaths;
function expandCSSPaths(path) {
    return ['.css', '.less', '.scss', '.sass'].map((ext) => {
        return `${path}${ext}`;
    });
}
exports.expandCSSPaths = expandCSSPaths;
