// @umijs/bundler-esbuild/dist/utils/postcssProcess.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.postcssProcess = void 0;
const postcss_1 = __importDefault(require("postcss"));
const getBrowserlist_1 = require("./getBrowserlist");
async function postcssProcess(config, css, path) {
    return await (0, postcss_1.default)([
        require('postcss-flexbugs-fixes'),
        require('postcss-preset-env')({
            browsers: (0, getBrowserlist_1.getBrowserlist)((config === null || config === void 0 ? void 0 : config.targets) || {}),
            autoprefixer: {
                flexbox: 'no-2009',
                ...config === null || config === void 0 ? void 0 : config.autoprefixer,
            },
            stage: 3,
        }),
    ].concat((config === null || config === void 0 ? void 0 : config.extraPostCSSPlugins) || [])).process(css, {
        from: path,
        to: path,
    });
}
exports.postcssProcess = postcssProcess;
