// @umijs/lint/dist/linter/base.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const child_process_1 = require("child_process");
const path_1 = __importDefault(require("path"));
/**
 * base linter
 */
class BaseLinter {
    constructor({ cwd }) {
        /**
         * linter package name
         */
        this.linter = '';
        /**
         * paths for linter
         */
        this.paths = {};
        this.paths.cwd = cwd;
    }
    /**
     * get bin file path for current linter
     */
    getBinPath() {
        try {
            const pkgPath = require.resolve(`${this.linter}/package.json`);
            const pkgContent = require(pkgPath);
            return path_1.default.resolve(path_1.default.dirname(pkgPath), pkgContent.bin[this.linter]);
        }
        catch (e) {
            throw new Error(`${this.linter} not found, please install it first.`);
        }
    }
    /**
     * get linter fork args
     */
    getRunArgs(args) {
        // not implemented
        args;
        return [];
    }
    /**
     * execute linter
     */
    run(args) {
        (0, child_process_1.fork)(this.getBinPath(), this.getRunArgs(args));
    }
}
exports.default = BaseLinter;
