// @umijs/preset-umi/dist/commands/generators/mock.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@umijs/core");
const path_1 = require("path");
const constants_1 = require("../../constants");
const utils_1 = require("./utils");
exports.default = (api) => {
    api.describe({
        key: 'generator:mock',
    });
    api.registerGenerator({
        key: 'mock',
        type: core_1.GeneratorType.generate,
        name: 'Generate mock',
        description: 'Generate mock boilerplate code',
        fn: async (opts) => {
            let [_, mockName] = opts.args._;
            const h = new utils_1.GeneratorHelper(api);
            mockName = await h.ensureVariableWithQuestion(mockName, {
                type: 'text',
                message: 'please input your mock file name',
                initial: 'mockName',
                format: utils_1.trim,
            });
            opts.generateFile({
                target: (0, path_1.join)(api.paths.cwd, 'mock', `${mockName}.ts`),
                baseDir: api.paths.cwd,
                path: (0, path_1.join)(constants_1.TEMPLATES_DIR, 'generate/mock.ts.tpl'),
                data: { mockName },
            });
        },
    });
};
