// @umijs/preset-umi/dist/commands/generators/prettier.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@umijs/core");
const utils_1 = require("@umijs/utils");
const fs_1 = require("fs");
const path_1 = require("path");
const utils_2 = require("./utils");
exports.default = (api) => {
    api.describe({
        key: 'generator:prettier',
    });
    api.registerGenerator({
        key: 'prettier',
        name: 'Enable Prettier',
        description: 'Setup Prettier Configurations',
        type: core_1.GeneratorType.enable,
        checkEnable: () => {
            // 存在 .prettierrc，不开启
            return !(0, fs_1.existsSync)((0, path_1.join)(api.cwd, '.prettierrc'));
        },
        disabledDescription: 'prettier has been enabled; You can remove `.prettierrc` to run this again to re-setup.',
        fn: async () => {
            const h = new utils_2.GeneratorHelper(api);
            h.addDevDeps({
                prettier: '^2',
                'prettier-plugin-organize-imports': '^2',
                'prettier-plugin-packagejson': '^2',
            });
            // 2、添加 .prettierrc 和 .prettierignore
            (0, fs_1.writeFileSync)((0, path_1.join)(api.cwd, '.prettierrc'), `
{
  "printWidth": 80,
  "singleQuote": true,
  "trailingComma": "all",
  "proseWrap": "never",
  "overrides": [{ "files": ".prettierrc", "options": { "parser": "json" } }],
  "plugins": ["prettier-plugin-organize-imports", "prettier-plugin-packagejson"]
}
`.trimLeft());
            utils_1.logger.info('Write .prettierrc');
            (0, fs_1.writeFileSync)((0, path_1.join)(api.cwd, '.prettierignore'), `
node_modules
.umi
.umi-production
`.trimLeft());
            utils_1.logger.info('Write .prettierignore');
            h.installDeps();
        },
    });
};
