// @umijs/preset-umi/dist/commands/generators/tailwindcss.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@umijs/core");
const utils_1 = require("@umijs/utils");
const fs_1 = require("fs");
const path_1 = require("path");
const utils_2 = require("./utils");
exports.default = (api) => {
    api.describe({
        key: 'generator:tailwindcss',
    });
    api.registerGenerator({
        key: 'tailwindcss',
        name: 'Enable Tailwind CSS',
        description: 'Setup Tailwind CSS configuration',
        type: core_1.GeneratorType.enable,
        checkEnable: () => {
            return !api.config.tailwindcss;
        },
        disabledDescription: () => `tailwindcss has been enabled; you can remove \`tailwindcss\` fields in ${api.appData.mainConfigFile} then run this to re-setup`,
        fn: async () => {
            const h = new utils_2.GeneratorHelper(api);
            h.addDevDeps({
                '@umijs/plugins': (0, utils_2.getUmiJsPlugin)(),
                tailwindcss: '^3',
            });
            h.setUmirc('tailwindcss', {});
            h.appendInternalPlugin('@umijs/plugins/dist/tailwindcss');
            utils_1.logger.info('Update .umirc.ts');
            (0, fs_1.writeFileSync)((0, path_1.join)(api.cwd, 'tailwind.config.js'), `
module.exports = {
  content: [
    './src/pages/**/*.tsx',
    './src/components/**.tsx',
    './src/layouts/**.tsx',
  ],
}
`.trimLeft());
            utils_1.logger.info('Write tailwind.config.js');
            (0, fs_1.writeFileSync)((0, path_1.join)(api.cwd, 'tailwind.css'), `
@tailwind base;
@tailwind components;
@tailwind utilities;
`.trimLeft());
            utils_1.logger.info('Write tailwind.css');
            h.installDeps();
        },
    });
};
