// @umijs/preset-umi/dist/commands/generators/tsconfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@umijs/core");
const utils_1 = require("@umijs/utils");
const fs_1 = require("fs");
const path_1 = require("path");
const utils_2 = require("./utils");
exports.default = (api) => {
    api.describe({
        key: 'generator:tsconfig',
    });
    api.registerGenerator({
        key: 'tsconfig',
        name: 'Enable Typescript',
        description: 'Setup tsconfig.json',
        type: core_1.GeneratorType.enable,
        checkEnable: () => {
            return !(0, fs_1.existsSync)((0, path_1.join)(api.paths.cwd, 'tsconfig.json'));
        },
        disabledDescription: 'tsconfig has been enabled; you can remove tsconfig.json then run this again to re-setup',
        fn: async () => {
            const h = new utils_2.GeneratorHelper(api);
            h.addDevDeps({
                typescript: '^4',
            });
            (0, fs_1.writeFileSync)((0, path_1.join)(api.cwd, 'tsconfig.json'), `
{
  "compilerOptions": {
    "target": "esnext",
    "module": "esnext",
    "moduleResolution": "node",
    "importHelpers": true,
    "jsx": "react",
    "esModuleInterop": true,
    "sourceMap": true,
    "baseUrl": ".",
    "strict": true,
    "paths": {
      "@/*": ["*"],
      "@@/*": [".umi/*"]
    },
    "allowSyntheticDefaultImports": true
  }
}
`.trimLeft());
            utils_1.logger.info('Write tsconfig.json');
            h.installDeps();
        },
    });
};
