// @umijs/utils/dist/register.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.restore = exports.clearFiles = exports.getFiles = exports.register = void 0;
const path_1 = require("path");
const pirates_1 = require("../compiled/pirates");
const COMPILE_EXTS = ['.ts', '.tsx', '.js', '.jsx'];
const HOOK_EXTS = [...COMPILE_EXTS, '.mjs'];
let registered = false;
let files = [];
let revert = () => { };
function transform(opts) {
    const { code, filename, implementor } = opts;
    files.push(filename);
    const ext = (0, path_1.extname)(filename);
    return implementor.transformSync(code, {
        loader: ext.slice(1),
        // consistent with `tsconfig.base.json`
        // https://github.com/umijs/umi-next/pull/729
        target: 'es2019',
        format: 'cjs',
    }).code;
}
function register(opts) {
    files = [];
    if (!registered) {
        revert = (0, pirates_1.addHook)((code, filename) => transform({ code, filename, implementor: opts.implementor }), {
            ext: opts.exts || HOOK_EXTS,
            ignoreNodeModules: true,
        });
        registered = true;
    }
}
exports.register = register;
function getFiles() {
    return files;
}
exports.getFiles = getFiles;
function clearFiles() {
    files = [];
}
exports.clearFiles = clearFiles;
function restore() {
    revert();
    registered = false;
}
exports.restore = restore;
